presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/WindowFilterPushDown.java

```diff
@@ -60,220 +64,154 @@ public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types
+                RowNumberNode rowNumberNode = (RowNumberNode) source;
+                int limit = Ints.checkedCast(node.getCount());
+                if (rowNumberNode.getMaxRowCountPerPartition().isPresent()) {
+                    limit = Math.min(rowNumberNode.getMaxRowCountPerPartition().get(), limit);
+                }
+                source = rowNumberNode.withLimit(Ints.checkedCast(limit));
+            }
+            else if (source instanceof WindowNode) {
+                WindowNode windowNode = (WindowNode) source;
+                if (canOptimizeWindowFunction(windowNode)) {
+                    // verify that unordered row_number window functions are replaced by RowNumberNode
+                    verify(!windowNode.getOrderBy().isEmpty());
+                    source = convertToTopNRowNumber(windowNode, Ints.checkedCast(node.getCount()));
+                }
             }
-
-            return context.defaultRewrite(node);
+            return context.replaceChildren(node, ImmutableList.of(source));
         }
 
         @Override
-        public PlanNode visitFilter(FilterNode node, RewriteContext<Constraint> context)
+        public PlanNode visitFilter(FilterNode node, RewriteContext<Void> context)
         {
-            PlanNode rewrittenSource = context.rewrite(node.getSource(), new Constraint(Optional.empty(), Optional.of(node.getPredicate())));
-            if (rewrittenSource != node.getSource()) {
-                if (rewrittenSource instanceof TopNRowNumberNode) {
-                    return rewrittenSource;
-                }
-                return new FilterNode(idAllocator.getNextId(), rewrittenSource, node.getPredicate());
-            }
-            return context.defaultRewrite(node);
-        }
-    }
+            PlanNode source = context.rewrite(node.getSource());
 
-    private static boolean isRowNumberSignature(Signature signature)
-    {
-        return signature.equals(ROW_NUMBER_SIGNATURE);
-    }
+            if (source instanceof RowNumberNode || source instanceof WindowNode) {
 
-    private static class Constraint
-    {
-        private final Optional<Integer> limit;
-        private final Optional<Expression> filterExpression;
+                TupleDomain<Symbol> tupleDomain = fromPredicate(metadata, session, node.getPredicate(), symbolAllocator.getTypes()).getTupleDomain();
 
-        private Constraint(Optional<Integer> limit, Optional<Expression> filterExpression)
-        {
-            this.limit = limit;
-            this.filterExpression = filterExpression;
-        }
-
-        public Optional<Integer> getLimit()
-        {
-            return limit;
+                if (source instanceof RowNumberNode) {
+                    Optional<Integer> highMarker = extractHighMarker(tupleDomain, ((RowNumberNode) source).getRowNumberSymbol());
+                    if (highMarker.isPresent()) {
+                        source = ((RowNumberNode) source).withLimit(highMarker.get());
```

Comment:
maybe we should just change the withLimit() here to be mergeLimit() and take care of it inside the function call.